### PR TITLE
xDSL/Vector: Add support for maskedstore, print and createmask op

### PR DESCRIPTION
**Pull request by meshtag**
This PR intends to extend Vector dialect by adding support for `vector.maskedstore`, `vector.print` and `vector.createmask` operations. The design is inspired by MLIR and I have tried to maintain semantic consistency. You can find corresponding MLIR semantics and verifiers for these operations here:

 - `vector.maskedstore`: [Doc Link](https://mlir.llvm.org/docs/Dialects/Vector/#vectormaskedstore-mlirvectormaskedstoreop), [Source code link](https://github.com/llvm/llvm-project/blob/8d92a8226eb71f0199f4cc83b961bf5440a1e503/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L1762)
 - `vector.print`: [Doc link](https://mlir.llvm.org/docs/Dialects/Vector/#vectorprint-mlirvectorprintop), [Source code link](https://github.com/llvm/llvm-project/blob/8d92a8226eb71f0199f4cc83b961bf5440a1e503/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L2408)
 - `vector.createmask`: [Doc link](https://mlir.llvm.org/docs/Dialects/Vector/#vectorcreate_mask-mlirvectorcreatemaskop), [Source code link](https://github.com/llvm/llvm-project/blob/8d92a8226eb71f0199f4cc83b961bf5440a1e503/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L2242)

Generated `.mlir` code from `.xdsl` testing file works with `llvm-project` commit hash: `74992f4a5bb79e2084abdef406ef2e5aa2024368`. 

Corresponding `.mlir` code can be generated by running: 
```
./xdsl/tools/xdsl-opt -t mlir -o tmp.mlir tests/filecheck/dialects/vector/vector_ops.xdsl
```

which can be further lowered to `llvm ir` by running: 
```
path-to-llvm-project/build/bin/mlir-opt --convert-scf-to-cf --convert-cf-to-llvm --convert-func-to-llvm --convert-arith-to-llvm --convert-memref-to-llvm --convert-vector-to-llvm --reconcile-unrealized-casts tmp.mlir | path-to-llvm-project/build/bin/mlir-translate --mlir-to-llvmir > tmp.ll
```

### CODE REVIEW

File: xdsl/dialects/vector.py
Comment (webmiche) [inline xdsl/dialects/vector.py:185]: Maybe report the types in this error message?

**Review by webmiche — COMMENTED** [review]
LGTM overall. Weirdly enough, I don't seem to be able to access the MLIR website, so I can't check their definitions exactly. Therefore, I would prefer to wait with merging until that is up again.

File: xdsl/dialects/vector.py
Comment (meshtag) [reply xdsl/dialects/vector.py]: For consistency, I'll create a separate PR to modify type check warnings in a similar fashion for other operations as well once this patch lands (I can do that in this PR itself but I think those changes will be unrelated to the PR/commit title and it might not be a good idea to fuse them here for a clean commit history).

File: xdsl/dialects/vector.py
Comment (webmiche) [inline xdsl/dialects/vector.py:195]: IMO, it would be cool to be able to encode this invariant as part of the annotation instead of in the verifier. That would make our annotation more declarative instead of the imperative verifier here.Though this might be more of a topic for a later, bigger xDSL PR. (@math-fehr)

File: xdsl/dialects/vector.py
Comment (webmiche) [inline xdsl/dialects/vector.py:213]: Please use `AnyAttr()` here. The `Attribute` annotation is correct, but feels a bit weird, IMO. AnyAttr nicely shows that this is just aynthing.

**Review by webmiche — APPROVED** [review]
LGTM, modulo one minor comment :)

File: xdsl/dialects/vector.py
Comment (meshtag) [reply xdsl/dialects/vector.py:195]: Yes, I agree. I was thinking of modelling at least `VectorOfRankAndType` (for verifying vector rank and type) and `VectorOfRank` (for verifying vector rank) verifier annotations for the `vector` dialect. I'll have a look at other dialects to see if they have repeated verifier mechanisms as well. Perhaps we should address all of them across all dialects in a separate PR.

File: xdsl/dialects/vector.py
Comment (superlopuh) [inline xdsl/dialects/vector.py:185]: Why not string comprehension here? I don't think we use much `str()` elsewhere in the code

Comment (meshtag) [reply xdsl/dialects/vector.py:185]: I used `str()` to print types nicely (like `!i32`) instead of `IntegerType(parameters=[IntAttr(data=32), SignednessAttr(data=<Signedness.SIGNLESS: 0>)], width=IntAttr(data=32), signedness=SignednessAttr(data=<Signedness.SIGNLESS: 0>))`. 

Do you think I should use something else for the same effect?

Comment (superlopuh) [reply xdsl/dialects/vector.py:185]: #395 Filed an issue, I think this can be cleaned up later in a consistent way

**Review by superlopuh — APPROVED** [review]

**Comment by meshtag** [comment]
> SOme tests need to be fixed as of the frozenlist removal, otherwise gtg

Ahh, I see. I wonder why only new tests failed in CI (I expect all of them to fail because of the frozenlist removal).

**Review by georgebisbas — CHANGES_REQUESTED** [review]
This PR has a lot of repeating code in tests. Can we do something to reduce this?
Otherwise GTG

**Comment by meshtag** [comment]
> This PR has a lot of repeating code in tests. Can we do something to reduce this?

I'll see if I can do something for this, I'll mostly try to modify all vector dialect tests in a single commit (in a separate PR, if that's fine).

**Review by georgebisbas — APPROVED** [review]

**Pull request merged**
